lib/contact.rb

```diff
@@ -88,6 +88,75 @@ def search_last_name(search_term)
     end
   end
 
+  def search_email(search_term)
+    sort_by_first_name
+    search = @contacts.select do |entry|
+      entry["email_address"].start_with?(search_term.downcase)
+    end
+
+    puts "Contacts with an email address starting with #{search_term.downcase}:"
+    puts ""
+    search.each do |entry|
+      entry.each do |key, value|
+        puts "#{key.to_s}: #{value}"
+      end
+      puts ""
+    end
+  end
+
+  def search_phone(search_term)
+    sort_by_first_name
+    search = @contacts.select do |entry|
+      entry["phone_number"].start_with?(search_term.downcase)
+    end
+
+    puts "Contacts with a phone number starting with #{search_term}:"
+    puts ""
+    search.each do |entry|
+      entry.each do |key, value|
+        puts "#{key.to_s}: #{value}"
+      end
+      puts ""
+    end
+  end
+
+  def delete_contact
+    puts ""
+    @contacts.each_with_index do |entry, index|
+      puts "#{index + 1}:"
+      entry.each do |key, value|
+        puts "#{key.to_s}: #{value}"
+      end
+      puts ""
+    end
+    puts "Please select which contact you wish to delete (select a number)"
+
+    selection = $stdin.gets.chomp.to_i
+
+    while (1..@contacts.length).to_a.include?(selection) == false
+      puts "You've entered an invalid selection. Please try again:"
+      selection = $stdin.gets.chomp.to_i
+    end
```

Comment:
This is more of a style thing but I quite like to put blocks of code like this in helper methods, for example you could write:

```ruby
def get_selection
    selection = $stdin.gets.chomp.to_i
    while (1..@contacts.length).to_a.include?(selection) == false
      puts "You've entered an invalid selection. Please try again:"
      selection = $stdin.gets.chomp.to_i
    end
    selection
end
```

And then you can use that in `delete_contact`

```ruby
selection = get_selection
```

A good practice is also to keep these helper methods private (so others classes that use `Contact` can't call them by accident). You can do this by declaring these methods near the bottom of the class under the `private` keyword, for example:

```ruby
class Contact
   def delete_contact
       ...
   end

   private

   def get_selection
       ...
   end
end
```

Let me know if that doesn't make sense!

Reply:
This is really helpful, thank you. I'll make this change now